File: python/twoSums.py
#twoSums using a hashMap
dict = {}
def hashMap(nums, target):
  for i, num in enumerate(nums):
    print(f' this is a list comp with string literals: {[(index, num) for index, num in enumerate(nums)]}')
    complement = target - num
    if complement in dict:
      return [dict[complement], i ]
    dict[num] = i

print(hashMap([3,7,4,10,2], 6))
print(dict)

# only need the below self if on leet code.

def twoSumTwo(self, nums, target):
  for num in range(len(nums)):
    for j in range(num + 1, len(nums)):
      if nums[num] + nums[j] == target:
        return [num, j]

File: javascript/twoSums.js
const twoSum = (nums, target) => {
  const hashMap = {};
  for(const [index, value] of nums.entries()){
    let complement = target - value
    if(complement in hashMap){
      return [hashMap[complement], index]
    }
    hashMap[value] = index
  }
}
// //Hash Map Solution
// Time Complexity: O(n)
// Each element is processed once, and hash map lookups/inserts are average O(1).

// Space Complexity: O(n)
// In the worst case, you store every element in the hash map.

// could also do the below response, which would be slower time complexity

const secondApporach =(nums, target) => {
  for(let i=0;i<nums.length;i++) {
    for(let j =i +1;j <nums.length; j++){
      if(nums[i] +nums[j]===target){
        return [i, j]
      }
    }
  }
}

// Nested Loops (for i, for j) Solution
// Time Complexity: O(n²)
// For each element (outer loop), the inner loop scans through the remaining elements.

// Space Complexity: O(1)
// No additional data structures are used aside from the return value.